x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/optimizer/OptimizerUtils.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ql.optimizer;
+
+public class OptimizerUtils {
+    private static final int HALF_FLOAT_MAX = 65504;
+
+    /**
+     * Check if an object  is a numerical value that fits into an integer.
+     */
+    public static boolean isInIntegerRange(Object num) {
```

Comment:
All these classes implement Number so this can be simplified to 
Number.longValue() < > Integer.MIN/MAX
However this will likely not maintain the original range so widening to long and double is necessary (which is what the JVM does behind the scene).

This is a typical Java gotcha hence why you need to check this works as expected.

Reply:
Ended up using big decimals to deal with the widening in a simple way.